### PR TITLE
feat(core): add method for requesting a lost block from a random session

### PR DESCRIPTION
**Pull request by lrubiorod**
Created a `RequestBlock` msg and handler in `session` actor to send a inventory_request_message with the `InventoryEntry::Block(Hash)` of the lost block.
Created a function to send an `Anycast<RequestBlock>` to `SessionsManager` actor.
Added updating `chain_info` for older blocks.
Added docs

### CODE REVIEW

File: core/src/actors/chain_manager/mod.rs [hunk 1]
```diff
@@ -315,6 +319,16 @@ impl ChainManager {
         });
     }
 
+    fn request_block(&mut self, block_hash: Hash) {
+        // Get SessionsManager's address
```
Comment (jrmoreau) [inline core/src/actors/chain_manager/mod.rs]:
```suggestion
        // Get SessionsManager address
```

File: core/src/actors/session/handlers.rs [hunk 2]
```diff
@@ -208,7 +208,20 @@ impl Handler<SendBlock> for Session {
             "Sending SendBlock message to peer at {:?}",
             self.remote_addr
         );
-        send_block_msg(self, &msg.block)
+        send_block_msg(self, msg.block)
+    }
+}
+
+/// Handler for SendBlock message (sent by other actors)
```
Comment (tmpolaczyk) [inline core/src/actors/session/handlers.rs]:
```suggestion
/// Handler for RequestBlock message (sent by other actors)
```

And add this message to the session documentation

File: core/src/actors/session/messages.rs [hunk 3]
```diff
@@ -45,3 +45,16 @@ impl fmt::Display for SendBlock {
         write!(f, "SendBlock")
     }
 }
+
+/// Message to send blocks through the network
```
Comment (tmpolaczyk) [inline core/src/actors/session/messages.rs]:
```suggestion
/// Message to request blocks through the network
```

File: core/src/actors/chain_manager/mod.rs [hunk 4]
```diff
@@ -439,7 +453,8 @@ impl ChainManager {
                         );
                     }
                 } else if !self.blocks.contains_key(&hash_prev_block) {
-                    //TODO Request this lost block to our peers
+                    self.request_block(hash_prev_block);
```
Comment (aesedepece) [inline core/src/actors/chain_manager/mod.rs]: I'd put a comment here explaining what's happening.

File: core/src/actors/session/handlers.rs [hunk 5]
```diff
@@ -542,16 +555,32 @@ fn send_item_msg(session: &mut Session, ctx: &mut Context<Session>, hash: &Hash)
...
+    let proof = block.proof;
+    let txns = block.txns;
     // Build Block msg
     let block_msg = WitnetMessage::build_block(block_header, proof, txns);
     // Send Block msg
     session.send_message(block_msg);
 }
 
+fn request_block_msg(session: &mut Session, block_entry: InventoryEntry) {
+    let mut inv_entries: Vec<InventoryEntry> = Vec::new();
```
Comment (aesedepece) [inline core/src/actors/session/handlers.rs]: These two lines can be blended into a single one:
```suggestion
    // Initialize a new inventory entries vector with the given block entry as its sole member
    let mut inv_entries: Vec<InventoryEntry> = vec![block_entry];
```

**Comment by tmpolaczyk** [comment]
I found a few bugs while trying a few nodes, but they were present before this PR:

* Since the blocks are now properly validated, we can't just start a node and expect it to work, because the previous block hash is 0000000..., which does not have a corresponding valid block. So we must somehow hard-code the genesis block as a valid block, at least for testing.

* Mining is broken, but it's a one line fix
<details>
  <summary>Show diff</summary>

```
diff --git a/core/src/actors/chain_manager/mining.rs b/core/src/actors/chain_manager/mining.rs
index 0de375b..33bd5ba 100644
--- a/core/src/actors/chain_manager/mining.rs
+++ b/core/src/actors/chain_manager/mining.rs
@@ -27,6 +27,11 @@ impl Handler<EpochNotification<MiningNotification>> for ChainManager {
     fn handle(&mut self, msg: EpochNotification<MiningNotification>, ctx: &mut Context<Self>) {
         debug!("Periodic epoch notification received {:?}", msg.checkpoint);
 
+        // Fix bug where this handler is executed before the other EpochNotification handler
+        // and it thinks the block it just mined is from the future
+        self.current_epoch = Some(msg.checkpoint);
+
+
         // Check eligibility
         // S(H(beacon))
         let mut beacon = match self.handle(GetHighestCheckpointBeacon, ctx) {
```
</details>

File: core/src/actors/chain_manager/mod.rs [hunk 6]
```diff
@@ -476,6 +493,22 @@ impl ChainManager {
                                 // Update UTXO set with older block transactions
                                 self.unspent_outputs_pool = utxo_set;
                                 self.transactions_pool = txn_mempool;
+
+                                // Update chain_info
```
Comment (tmpolaczyk) [inline core/src/actors/chain_manager/mod.rs]: Why is `chain_info` updated here? The block is still only a candidate, right?

**Pull request merged**